포커 딜러와 게임 테스트

TTOzzi opened the pull request:
테스트 코드를 최대한 한 기능만 테스트하도록 해주고 싶었지만 카드를 나눠주는 것을 테스트하기 위해선 플레이어가 필요하고 손패를 초기화하는 것을 테스트하기 위해서는 플레이어에게 카드를 나눠 준 뒤 테스트가 이루어져야 합니다.
이를 해결하기 위해 플레이어를 직접 추가해주고, 카드를 나눠주는 것도 수동으로 해주는 것도 생각해봤지만, 어차피 함수 안의 코드를 그대로 끄집어내어 추가해주는 거라 의미가 없다고 생각해서 그대로 뒀습니다.
뭔가 다른 좋은 방법이 있을까요..?

TTOzzi committed:
Player, Dealer 클래스 생성

플레이어 번호, 핸드, 딜러 핸드 추가
아직 별도 동작을 구현해주진 않음

TTOzzi committed:
FiveCardStud 클래스 생성

딜러 변수와 참여한 플레이어의 수를 담을 변수 추가

TTOzzi committed:
player 클래스 number 삭제

참가자 번호가 순서를 표시하는것 이외엔 사용되지 않아 배열의 index 로 대체할 수 있을 것 같아 삭제해줌

TTOzzi committed:
addPlayers() 함수 구현

처음 생성할 때 설정해준 인원수만큼 플레이어를 추가하고 딜러도 플레이어로 추가해줌
딜러도 게임플레이를 하므로 딜러와 플레이어를 같은 배열에 담기위해 딜러가 플레이어를 상속받게 해줌

TTOzzi committed:
addPlayer() 테스트코드 작성

플레이어와 딜러 추가 후 players 배열 갯수 확인

TTOzzi committed:
Dealer 클래스 drawCard() 구현

덱에서 카드를 뽑아 drawnCard 에 저장해줌

TTOzzi committed:
Dealer 클래스 drawCard() 테스트 코드 작성

카드를 뽑은 후 덱의 카드 수, 뽑은 카드를 잘 들고있는지 확인

TTOzzi committed:
FiveCardStud 클래스 카드배분(distributeCards) 구현

게임 내 각 플레이어들에게 딜러가 카드를 한장씩 뽑아 나눠줌

TTOzzi committed:
FiveCardStud 클래스 distributeCards 테스트 코드 구현

카드를 나눠준 뒤 각 플레이어의 핸드에 있는 카드 수 확인
하나의 기능만 테스트하고싶었지만 플레이어를 추가하지않고는 테스트가 불가능하여 플레이어를 먼저 추가해줌

TTOzzi committed:
FiveCardStud 클래스 run() 구현

플레이어를 추가해주고 각 플레이어에게 1장씩 5번 나눠줌

TTOzzi committed:
FiveCardStud 클래스 run() 테스트코드 작성

run() 실행 후 각 플레이어가 5장의 카드를 갖고있는지 확인

TTOzzi committed:
FiveCardStud 클래스를 PockerGame 클래스로 변경

7카드스터드 방식과 5카드스터드 방식의 차이점이 카드의 갯수뿐인듯 하여 gameType 변수를 만들어 7카드스터드와 5카드스터드를 구분해주도록 함

TTOzzi committed:
불필요 코드 제거

Deck 에 있던 print() 제거

TTOzzi committed:
Player 클래스 initHand() 구현

다음 게임을 시작하기 위한 손패 초기화 함수

TTOzzi committed:
테스트코드 폴더 명 변경

FiveCardStudTest 를 PokerGameTest 로 변경

TTOzzi committed:
게임 재시작을 위한 PokerGame 클래스 initHand(), play(), run() 구현, 테스트코드 작성

initHand() - 플레이어들의 핸드를 초기화
play() - 플레이어들의 핸드 초기화 후 카드를 게임타입에 맞게 나눠줌
run() - 카드가 부족할때까지 반복해서 게임을 진행함.

TTOzzi committed:
Dealer 클래스 shuffle() 추가, 게임시작 전 덱을 섞도록 수정

딜러가 덱을 섞어주도록 함

TTOzzi committed:
PokerGame 클래스 resumable 추가

게임 재시작이 가능한지 남은 카드수와 게임에 필요한 카드 수를 비교하여 알려줌

TTOzzi committed:
PokerGame 클래스 GameType 추가

gameType 의 타입을 Int 에서 GameType 으로 변경
enum 을 만들어 5카드, 7카드를 제외한 다른 선택지를 없앰

TTOzzi committed:
변경된 코드에 맞게 테스트코드 수정

TTOzzi committed:
ReadMe 업데이트

코드 변경사항, 출력결과 기록

godrm commented:
분량이 많아서 조금 천천히 보고 피드백 남길게요.

TTOzzi commented:
넵

godrm commented on CardGameApp/CardGameApp/Dealer.swift:
```diff
@@ -0,0 +1,22 @@
+    
+    func drawCard() {
+        drawnCard = deck.removeOne()
+    }
```
속성은 private var를 우선 고려하는 게 좋습니다. 
drawCard()는 그냥 return 해줘도 될 것 같은데 왜 속성에 넣고 그걸 참조해야 하는걸까요? 
이 함수가 정상적으로 동작했는지 바로 알려주는 게 좋지 않을까요?

TTOzzi replied:
현실에서 딜러가 카드를 뽑아서 갖고있다가 주는것을 표현하려고 했었는데 수정하겠습니다.

godrm commented on CardGameApp/CardGameApp/Player.swift:
```diff
@@ -0,0 +1,17 @@
+    
+    func initHand() {
+        hand = [Card]()
+    }
```
속성은 private var을 우선적으로 고려하세요
객체 외부에서 값을 바꿀 수 있으면 안 됩니다. 
initHand() 보다는 init()을 그냥 사용하게 좋을 것 같습니다. 
내부에서 어떤 작업을 하는지 명시할 필요는 없습니다. 
그리고 위에 리터럴하게 빈 배열을 만들었는데 또 넣을 필요가 있나요?

TTOzzi replied:
게임이 한번 진행된 후 핸드에 있는 카드들을 없애주기 위해 한 번 더 넣어주었었습니다.
init 으로 한 번에 처리하도록 수정하겠습니다.

godrm replied:
기존에 있는 카드를 제거하는 것은 init-으로 시작하면 안 되고 다른 이름이어야 하지 않을까요?

TTOzzi replied:
손 패를 초기화한다는 의미로 initHand() 라고 지어줬었는데 init 을 안 쓰는 이름으로 한다면 discard() 로 바꿔주면 될 것 같아요.
게임을 재시작할 때 기존에 있는 카드를 제거하는 것을 init() 으로 처리해주려고 했는데 이미 플레이어가 생성된 상태에선 init 을 호출하지 못하는 것 같아 기존 카드를 버리는 메서드는 따로 있어야 할 것 같아요.

godrm replied:
네. init-으로 시작하는 것은 생성자라고 인지하게 됩니다. 그래서 규칙을 잘 알아야 하는거죠. 
이미 인스턴스가 만들어진 상태에서는 다른 메소드로 초기화를 하는 게 좋겠죠. 
근데 이 함수가 public 으로 오픈되어야 하는지도 다시 한 번 살펴보세요.

TTOzzi replied:
이 함수가 사용되는 곳이 PokerGame 클래스에서 다음 게임이 시작될 때인데 private 으로 설정해주면 PokerGame 에서 못 써서 public 으로 둬야 할 것 같은데 다른 방법이 있나요?

godrm replied:
그러면 할 수 없겠네요 :)

godrm commented on CardGameApp/CardGameApp/PokerGame.swift:
```diff
@@ -0,0 +1,67 @@
+class PokerGame {
+    var dealer = Dealer()
+    var players = [Player]()
+    var numberOfPlayers: Int
```
이 변수는 위에 배열 크기와 항상 같지 않을까요?

TTOzzi replied:
numberOfPlayers 는 처음 게임을 생성할 때 참가자의 수라 배열에 딜러를 추가해주면 + 1 이 되어 따로 나눠주었었는데 생각해보니 더 좋은 방법이 있을 것 같습니다. 수정해보겠습니다.

godrm commented on CardGameApp/CardGameApp/PokerGame.swift:
```diff
@@ -0,0 +1,67 @@
+    
+    enum GameType: Int {
+        case fiveCardStud = 5
+        case sevenCardStud = 7
```
nested 타입 선언은 사용한 속성보다 윗쪽에 작성하는 게 좋습니다

godrm replied:
개인적으로는 nested 타입은 모든 변수를 선언하기 이전에 선언하는 편입니다. 참고하세요.

TTOzzi replied:
앗 알겠습니다

godrm commented on CardGameApp/CardGameApp/PokerGame.swift:
```diff
@@ -0,0 +1,67 @@
+            guard let drawnCard = dealer.drawnCard else {
+                return
+            }
+            $0.hand.append(drawnCard)
```
이렇게 직접적으로 외부에서 내부에 값을 추가하기 보다는 메소드로 전달하도록 구현하는 게 좋습니다. 객체를 보호할 수 있으니까요

godrm commented on CardGameApp/CardGameApp/PokerGame.swift:
```diff
@@ -0,0 +1,67 @@
+        players.forEach {
+            $0.initHand()
+        }
+    }
```
이 메소드가 명시적으로 있어야 할 이유를 모르겠습니다. 
내부적으로 리셋이 필요한 경우만 호출해도 되지 않을까요?

godrm commented on CardGameApp/PokerGameTest/PokerGameTest.swift:
```diff
@@ -0,0 +1,63 @@
+        XCTAssertEqual(fiveCardStud.players.count, 3)
+        sevenCardStud.addPlayers()
+        XCTAssertEqual(sevenCardStud.players.count, 4)
+    }
```
같은 기능이더라도 조건이 다르면 태스트 함수를 분리하세요
왜냐하면 위에 assert가 실패한 건지 아래거가 실패한 건지 
함수 이름만 보고 알 수 있어야 합니다. 
지금은 7카드일 때 에러인지 5카드일 때 에러인지 구분할 수 없네요

godrm reviewed (commented):
다음 피드백들을 확인하세요
모든 파일에 적용되어야 하는 피드백도 있습니다.
PR 열어놓을테니 수정하고 넘어가도록 하겠습니다.

TTOzzi commented:
수정하고 코멘트 남기겠습니다!

TTOzzi committed:
Dealer 클래스 drawCard() 수정

drawnCard 삭제, 카드를 바로 리턴하도록 수정
관련 코드 수정

TTOzzi committed:
PokerGame 클래스 init() 수정

numberOfPlayers, addPlayers 를 없애고 생성과 동시에 플레이어들을 추가해주도록 수정

TTOzzi committed:
Player 클래스 receive(card:) 추가

TTOzzi committed:
Player 클래스 init() 추가, initHand() 를 discard() 로 변경

TTOzzi committed:
PokerGame 클래스 initHand() 삭제

play() 에 코드를 포함시킴

TTOzzi committed:
PokerGame 클래스 run() 과 play() 통합

addPlayers 를 없애고, initHand 를 play 와 합쳐줘서 run 의 역할이 단순해짐.
play 와 합쳐줌

TTOzzi committed:
PokerGame 테스트코드 5카드스터드, 7카드스터드 분리

TTOzzi committed:
전체 코드 접근제어 설정

Dealer 클래스 deck 을 pirvate 으로 설정, 덱의 카드의 갯수를 반환하는 연산프로퍼티 생성
Player 클래스 hand 를 pirvate 으로 설정, 핸드의 카드의 갯수를 반환하는 연산프로퍼티 생성
PokerGame 클래스의 dealer, gameType, resumable private 으로 설정

TTOzzi committed:
수정된 코드에 맞게 테스트코드 수정

TTOzzi commented:
피드백들 수정 완료했습니다.
속성들을 private 로 설정해주다 보니 테스트가 불가능하여 Player 클래스에 실제 게임플레이 로직엔 필요 없는 handCount 연산프로퍼티를 만들어주었는데 이게 좋은 방법인지 모르겠습니다.

godrm commented on CardGameApp/CardGameApp/PokerGame.swift:
```diff
@@ -9,25 +9,20 @@
+    }
     
-    func addPlayers() {
+    init(gameType: GameType, numberOfPlayers: Int) {
```
numberOfPlayers 도 GameType 처럼 enum으로 선언하는 건 어떤가요? 
Int 나 Double 같은 타입은 모든 숫자를 다 표현하는 타입입니다. 
게임을 초기화할 때 numberOfPlayers 값을 100 이나 1000 보다 더 큰 값을 넣으면 가능한가요?

TTOzzi replied:
넵 이것도 바꿔서 선택지를 제한하겠습니다.

godrm commented:
수고하셨습니다.

TTOzzi commented:
감사합니다:grinning:
